ql/indexes/equityindex.cpp

```diff
@@ -0,0 +1,99 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+/*
+ Copyright (C) 2023 Marcin Rybacki
+
+ This file is part of QuantLib, a free-software/open-source library
+ for financial quantitative analysts and developers - http://quantlib.org/
+
+ QuantLib is free software: you can redistribute it and/or modify it
+ under the terms of the QuantLib license.  You should have received a
+ copy of the license along with this program; if not, please email
+ <quantlib-dev@lists.sf.net>. The license is also available online at
+ <http://quantlib.org/license.shtml>.
+
+ This program is distributed in the hope that it will be useful, but WITHOUT
+ ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ FOR A PARTICULAR PURPOSE.  See the license for more details.
+*/
+
+#include <ql/indexes/equityindex.hpp>
+#include <ql/settings.hpp>
+#include <utility>
+
+namespace QuantLib {
+
+    namespace {
+        Real resolveSpot(const Handle<Quote>& spot, Real todaysFixing) {
+            QL_REQUIRE(!spot.empty() || todaysFixing != Null<Real>(),
+                       "Cannot forecast equity index, missing both spot and historical index");
+            return spot.empty() ? todaysFixing : spot->value();
+        }
+    }
+
+    EquityIndex::EquityIndex(std::string name,
+                             Calendar fixingCalendar,
+                             Handle<YieldTermStructure> interest,
+                             Handle<YieldTermStructure> dividend,
+                             Handle<Quote> spot)
+    : name_(std::move(name)), settlementCalendar_(std::move(fixingCalendar)),
+      interest_(std::move(interest)), dividend_(std::move(dividend)), spot_(std::move(spot)) {
+
+        registerWith(interest_);
+        registerWith(dividend_);
+        registerWith(spot_);
+        registerWith(Settings::instance().evaluationDate());
+        registerWith(IndexManager::instance().notifier(EquityIndex::name()));
+    }
+
+    Real EquityIndex::fixing(const Date& fixingDate, bool forecastTodaysFixing) const {
+
+        QL_REQUIRE(isValidFixingDate(fixingDate), "Fixing date " << fixingDate << " is not valid");
+
+        Date today = Settings::instance().evaluationDate();
+
+        if (fixingDate > today || (fixingDate == today && forecastTodaysFixing))
+            return forecastFixing(fixingDate);
+
+        Real result = pastFixing(fixingDate);
+
+        if (result != Null<Real>())
+            // if historical fixing is present use it
+            return result;
```

Comment:
In the case of fixingDate = today, here you're preferring the stored fixing (if any) over the value in the spot handle.
In `resolveSpot` you're preferring the handle instead.  Is there any particular reasoning behind this?